tuning/tune.py

```diff
@@ -765,111 +904,52 @@ def tune(
-        K1 = product(val if dim == "k" else 1 for dim, val in zip(rhs_dims, RHS))
-        assert K0 == K1
-        tune_logger.debug(f"Equivalent matmul shape: [{M}, {N}, {K0}]")
-        problem_size = ProblemSize(
-            M, N, K0, *get_bit_width(), dispatch_kind=walk_result.dispatch_kind
+        get_shapes_fn = lambda template: get_shapes_contract(
+            template, lhs_dims, rhs_dims
+        )
+        apply_params_fn = lambda ps, template, config: (
+            apply_params_contract(ps, tile_dims, template, config),
+            "",
         )
-
-        for i, config in enumerate(generate_solutions(problem_size)):
-            if i >= limit:
-                break
-            tune_logger.info(f"Solution #{i+1}: {config}")
-            configs.append(config)
-            new_mlir = apply_params_contract(
-                LHS, RHS, RES, tile_dims, mlir_template, config
-            )
-
-            with open(path.join(output, f"{i+1}.mlir"), "w") as f:
-                f.write(new_mlir)
     elif walk_result.dispatch_kind == DispatchKind.batch_matmul:
-        LHS, RHS, RES = get_shapes_batch_matmul(mlir_template)
-        assert len(LHS) == len(lhs_dims)
-        assert len(RHS) == len(rhs_dims)
-        B = product(val if dim == "b" else 1 for dim, val in zip(lhs_dims, LHS))
-        B0 = product(val if dim == "b" else 1 for dim, val in zip(lhs_dims, RHS))
-        B1 = product(val if dim == "b" else 1 for dim, val in zip(lhs_dims, RES))
-        M = product(val if dim == "m" else 1 for dim, val in zip(lhs_dims, LHS))
-        N = product(val if dim == "n" else 1 for dim, val in zip(rhs_dims, RHS))
-        K0 = product(val if dim == "k" else 1 for dim, val in zip(lhs_dims, LHS))
-        K1 = product(val if dim == "k" else 1 for dim, val in zip(rhs_dims, RHS))
-        assert B == B0 and B == B1
-        assert K0 == K1
-        tune_logger.debug(f"Batch matmul shape: {B}x[{M}, {N}, {K0}]")
-        problem_size = ProblemSize(
-            M, N, K0, *get_bit_width(), dispatch_kind=walk_result.dispatch_kind
+        get_shapes_fn = lambda template: get_shapes_batch_matmul(
+            template, lhs_dims, rhs_dims
+        )
+        apply_params_fn = lambda _ps, template, config: apply_params_batch_matmul(
+            template, lhs_dims, rhs_dims, config
         )
-
-        for i, config in enumerate(generate_solutions(problem_size)):
-            if i >= limit:
-                break
-            tune_logger.info(f"Solution #{i+1}: {config}")
-            configs.append(config)
-            new_mlir, embeddable_tuning = apply_params_batch_matmul(
-                LHS, RHS, RES, B, M, N, K0, tile_dims, mlir_template, config
-            )
-
-            with open(path.join(output, f"{i+1}.mlir"), "w") as f:
-                f.write(new_mlir)
-            with open(path.join(output, f"{i+1}_config.mlir"), "w") as f:
-                f.write(embeddable_tuning)
     else:
         assert False
```

Comment:
maybe give more error details
```suggestion
        raise ValueError(f"Unknown dispatch kind: {dispatch_kind}")
```